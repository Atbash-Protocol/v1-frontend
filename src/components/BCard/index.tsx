import React from 'react';

import { Box, Typography, Zoom } from '@mui/material';

import { theme } from 'constants/theme';

interface BCardProps {
    title: string | JSX.Element;
    children: React.ReactNode;
    zoom: boolean;
    className?: string;
}

export const BCard = ({ title, children, zoom, className }: BCardProps) => {
    const renderTitle = React.isValidElement(title) ? (
        { title }
    ) : (
        <Typography variant="h4" sx={{ textTransform: 'uppercase' }}>
            {title}
        </Typography>
    );

    const card = (
        <Box
            className={className}
            sx={{
                padding: theme.spacing(4),
                [theme.breakpoints.up('xs')]: {
                    margin: '.5rem',
                },
                [theme.breakpoints.down('xs')]: {
                    margin: '1rem',
                },
                '@supports (-webkit-backdrop-filter: none) or (backdrop-filter: none)': {
                    background: theme.palette.cardBackground.main,
                    backdropFilter: 'blur(100px)',
                },

                '@supports not ((-webkit-backdrop-filter: none) or (backdrop-filter: none))': {
                    background: theme.palette.cardBackground.dark,
                    backdropFilter: 'blur(100px)',
                },
            }}
        >
            <Box sx={{ color: theme.palette.primary.main }}>
                <>{renderTitle}</>
            </Box>
            {children}
        </Box>
    );

    return zoom ? (
        <>
            <Zoom in={false}>
                <>{card}</>
            </Zoom>
        </>
    ) : (
        card
    );
};
