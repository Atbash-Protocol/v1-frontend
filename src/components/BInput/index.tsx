import { ChangeEvent, SyntheticEvent, useState } from 'react';

import { OutlinedInput, InputAdornment, Box, Typography, SxProps, Theme } from '@mui/material';
import { useTranslation } from 'react-i18next';

import { theme } from 'constants/theme';

interface BInputProps {
    name: string;
    defaultValue: string;
    onChange: (value: Record<string, string>) => void;
    maxValue?: string;
    placeholder?: string;
    endAdornmentLabel?: string;
    extraSxProps?: SxProps<Theme> | undefined;
}

const BInput = ({ name, defaultValue, maxValue, onChange, endAdornmentLabel, placeholder, extraSxProps }: BInputProps) => {
    const { t } = useTranslation();
    const [value, setValue] = useState(defaultValue);

    const handleClickMaxValue = (event: SyntheticEvent<HTMLDivElement>) => {
        event.preventDefault();
        setValue(maxValue ?? defaultValue);
    };

    const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
        event.preventDefault();
        const inputValue = event.target.value;

        setValue(inputValue);
        onChange({ [name]: inputValue });
    };

    return (
        <OutlinedInput
            sx={{
                color: theme.palette.primary.main,
                border: '1px solid',
                borderColor: theme.palette.primary.main,
                borderRadius: 0,
                outlineColor: theme.palette.primary.main,
                width: '100%',
                ...(extraSxProps ?? {}),
            }}
            name={name}
            type="number"
            placeholder={placeholder}
            value={value}
            onChange={handleInputChange}
            inputProps={{ inputMode: 'numeric', pattern: '^[0-9]*[.,]?[0-9]*$' }}
            endAdornment={
                <InputAdornment position="end">
                    <Box sx={{ color: theme.palette.primary.main, textTransform: 'uppercase', cursor: 'pointer' }} onClick={handleClickMaxValue}>
                        <Typography>{endAdornmentLabel ?? t('Max')}</Typography>
                    </Box>
                </InputAdornment>
            }
        />
    );
};

export default BInput;
