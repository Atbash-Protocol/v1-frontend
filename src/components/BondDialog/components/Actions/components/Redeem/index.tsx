import { Box } from '@mui/material';

export const Redeem = ({ bondID }: { bondID: string }) => {
    return <Box>Redeem {bondID}</Box>;
};
