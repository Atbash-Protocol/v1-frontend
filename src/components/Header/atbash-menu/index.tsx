import { useEffect, useState } from 'react';
import { TOKEN_DECIMALS, DEFAULT_NETWORK, getAddressesAsync, IAddresses, Networks } from '../../../constants';
import { useSelector } from 'react-redux';
import { Link, Fade, Popper } from '@material-ui/core';
import './atbash-menu.scss';
import { IReduxState } from '../../../store/slices/state.interface';
import { getTokenUrl } from '../../../helpers';

import { useTranslation } from 'react-i18next';

const addTokenToWallet = (tokenSymbol: string, tokenAddress: string) => async () => {
    const tokenImage = getTokenUrl(tokenSymbol.toLowerCase());

    if (window.ethereum) {
        try {
            await window.ethereum.request({
                method: 'wallet_watchAsset',
                params: {
                    type: 'ERC20',
                    options: {
                        address: tokenAddress,
                        symbol: tokenSymbol,
                        decimals: TOKEN_DECIMALS,
                        image: tokenImage,
                    },
                },
            });
        } catch (error) {
            console.log(error);
        }
    }
};

function AtbashMenu() {
    const { t } = useTranslation();

    const [anchorEl, setAnchorEl] = useState(null);
    const isEthereumAPIAvailable = window.ethereum;
    const [chainArgument, setChainArgument] = useState('');

    const networkID = useSelector<IReduxState, number>(state => {
        return (state.app && state.app.networkID) || DEFAULT_NETWORK;
    });

    const initialState: IAddresses = {
        BASH_ADDRESS: '',
        BASH_BONDING_CALC_ADDRESS: '',
        BASH_DAI_BOND_ADDRESS: '',
        BASH_DAI_LP_ADDRESS: '',
        DAI_ADDRESS: '',
        DAI_BOND_ADDRESS: '',
        DAO_ADDRESS: '',
        REDEEM_ADDRESS: '',
        SBASH_ADDRESS: '',
        STAKING_ADDRESS: '',
        STAKING_HELPER_ADDRESS: '',
        TREASURY_ADDRESS: '',
        WSBASH_ADDRESS: '',
        ABASH_ADDRESS: '',
        PRESALE_ADDRESS: '',
        PRESALE_REDEMPTION_ADDRESS: '',
    };
    const [addresses, setAddresses] = useState<IAddresses>(initialState);
    const loadAddresses = async () => {
        const addresses = await getAddressesAsync(networkID);
        setAddresses(addresses);
        if (networkID == Networks.RINKEBY) setChainArgument('&chain=rinkeby');
    };
    useEffect(() => {
        loadAddresses();
    }, [networkID]);

    const handleClick = (event: any) => {
        setAnchorEl(anchorEl ? null : event.currentTarget);
    };

    const open = Boolean(anchorEl);

    return (
        <div className="bash-menu-root" onMouseEnter={e => handleClick(e)} onMouseLeave={e => handleClick(e)}>
            <div className="bash-menu-btn">
                <p>{t('BuyBASH')}</p>
            </div>

            <Popper className="bash-menu-popper" open={open} anchorEl={anchorEl} transition>
                {({ TransitionProps }) => (
                    <Fade {...TransitionProps} timeout={200}>
                        <div className="tooltip">
                            <Link
                                className="tooltip-item"
                                href={`https://app.uniswap.org/#/swap?&inputCurrency=${addresses.DAI_ADDRESS}&outputCurrency=${addresses.BASH_ADDRESS}${chainArgument}`}
                                target="_blank"
                            >
                                <p>{t('BuyOnUniswap')}</p>
                            </Link>

                            {isEthereumAPIAvailable && (
                                <div className="add-tokens">
                                    <div className="divider" />
                                    <p className="add-tokens-title">{t('AddTokenToWallet')}</p>
                                    <div className="divider" />
                                    <div className="tooltip-item" onClick={addTokenToWallet('BASH', addresses.BASH_ADDRESS)}>
                                        <p>↑ BASH</p>
                                    </div>
                                    <div className="tooltip-item" onClick={addTokenToWallet('sBASH', addresses.SBASH_ADDRESS)}>
                                        <p>↑ sBASH</p>
                                    </div>
                                </div>
                            )}
                        </div>
                    </Fade>
                )}
            </Popper>
        </div>
    );
}

export default AtbashMenu;
