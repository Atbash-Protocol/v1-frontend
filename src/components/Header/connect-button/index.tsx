import { useCallback, useContext, useEffect, useState } from 'react';

import { Box, Button } from '@mui/material';
import { useTranslation } from 'react-i18next';
import { useSelector } from 'react-redux';

import { theme } from 'constants/theme';
import { Web3Context } from 'contexts/web3/web3.context';
import { useSignerConnected, useGoodNetworkCheck } from 'contexts/web3/web3.hooks';
import { selectTransactionPending } from 'store/modules/transactions/transactions.selectors';

function ConnectMenu() {
    const { t } = useTranslation();
    const { memoConnect, memoDisconnect, state } = useContext(Web3Context);

    const isUserSigned = useSignerConnected();

    const isUserOnGoodNetwork = useGoodNetworkCheck();

    const isOneTransactionPending = useSelector(selectTransactionPending);

    const handleButtonClick = useCallback(() => {
        return isUserSigned ? memoDisconnect(state.signer) : memoConnect();
    }, [isUserSigned, memoConnect, memoDisconnect]);

    const [buttonText, setButtonText] = useState(t('ConnectWallet'));

    useEffect(() => {
        if (isUserSigned) {
            if (!isUserOnGoodNetwork) {
                setButtonText(t('WrongNetwork'));
            } else if (isOneTransactionPending) {
                setButtonText(t('CountPending', { count: 1 })); // Ususally user can't have more than 1
            } else {
                setButtonText(t('Disconnect'));
            }
        }
    }, [isUserSigned, isOneTransactionPending, isUserOnGoodNetwork]);

    return (
        <Box>
            <Button
                sx={{
                    background: isUserSigned && !isUserOnGoodNetwork ? 'red !important' : 'rgba(255, 255, 255, 0.9)',
                    boxShadow: '0px 0px 10px rgba(44, 39, 109, 0.1)',
                    padding: theme.spacing(1),
                }}
                onClick={handleButtonClick}
            >
                <>{buttonText}</>
            </Button>
        </Box>
    );
}

export default ConnectMenu;
