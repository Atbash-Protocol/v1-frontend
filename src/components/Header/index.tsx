import { makeStyles } from "@material-ui/core/styles";
import useMediaQuery from "@material-ui/core/useMediaQuery";
// import MenuIcon from "../../assets/icons/hamburger.svg";
import MenuIcon from "@mui/icons-material/Menu";

import AtbashMenu from "./atbash-menu";
import ConnectButton from "./connect-button";
import "./header.scss";
import { DRAWER_WIDTH, TRANSITION_DURATION } from "constants/styles";
import { AppBar, Box, Button, IconButton, Toolbar, Typography } from "@mui/material";
import { theme } from "constants/theme";

interface IHeader {
    handleDrawerToggle: () => void;
    isSmallScreen: boolean;
}

function Header({ handleDrawerToggle, isSmallScreen }: IHeader) {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <AppBar
                color="transparent"
                position="sticky"
                sx={{
                    background: "transparent",
                    border: "none",
                    boxShadow: "none",
                    position: "relative",
                    zIndex: theme.zIndex.drawer + 1,
                }}
            >
                <Toolbar
                    sx={{
                        background: "transparent",
                    }}
                >
                    {isSmallScreen && (
                        <IconButton onClick={handleDrawerToggle} size="large" edge="start" color="inherit" aria-label="menu" sx={{ mr: 2 }}>
                            <MenuIcon />
                        </IconButton>
                    )}
                    <Box sx={{ flexGrow: 1 }}></Box>
                    <AtbashMenu />
                    <ConnectButton />
                </Toolbar>
            </AppBar>
        </Box>
        // <div className={`${classes.topBar} ${!drawe && classes.topBarShift}`}>
        //     <AppBar position="sticky" className={classes.appBar} elevation={0}>
        //         <Toolbar className="dapp-topbar">
        //             <div onClick={handleDrawerToggle} className="dapp-topbar-slider-btn">
        //                 <img src={MenuIcon} alt="" />
        //             </div>
        //             <div className="dapp-topbar-btns-wrap">
        //                 {!isVerySmallScreen && <AtbashMenu />}
        //                 <ConnectButton />
        //             </div>
        //         </Toolbar>
        //     </AppBar>
        // </div>
    );
}

export default Header;
