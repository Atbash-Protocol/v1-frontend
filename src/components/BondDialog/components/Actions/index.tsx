import { ReactNode, SyntheticEvent, useState } from 'react';

import { Box, Tabs, Tab } from '@mui/material';
import { useTranslation } from 'react-i18next';

import { Mint } from './components/Mint';
import { Redeem } from './components/Redeem';

enum PanelEnum {
    MINT = 0,
    REDEEM = 1,
}

interface TabPanelProps {
    children?: ReactNode;
    index: number;
    value: number;
}

function TabPanel(props: TabPanelProps) {
    const { children, value, index, ...other } = props;

    return (
        <div role="tabpanel" hidden={value !== index} id={`stake-tabpanel-${index}`} aria-labelledby={`stake-tab-${index}`} {...other}>
            <Box sx={{ p: 3, color: 'red' }}>{children}</Box>
        </div>
    );
}

function a11yProps(index: number) {
    return {
        id: `stake-tab-${index}`,
        'aria-controls': `stake-tabpanel-${index}`,
        sx: {
            color: 'white',
        },
    };
}

export const Actions = ({ bondID }: { bondID: string }) => {
    const [value, setValue] = useState(PanelEnum.MINT);
    const { t } = useTranslation();

    const handleChange = (event: SyntheticEvent, newValue: number) => {
        event.preventDefault();
        setValue(newValue || 0);
    };

    return (
        <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
            <Tabs centered value={value} onChange={handleChange} aria-label="basic tabs example">
                <Tab value={0} label={<>{t('mint:Mint')}</>} {...a11yProps(0)} />
                <Tab value={1} label={<>{t('mint:Redeem')}</>} {...a11yProps(1)} />
            </Tabs>
            <TabPanel value={value} index={0}>
                <Mint bondID={bondID} />
            </TabPanel>
            <TabPanel value={value} index={1}>
                <Redeem bondID={bondID} />
            </TabPanel>
        </Box>
    );
};
