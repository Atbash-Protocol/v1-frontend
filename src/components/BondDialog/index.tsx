import { Box, Dialog, DialogContent, DialogTitle, Grid, Typography } from '@mui/material';
import { t } from 'i18next';
import { useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { BMultiTabs } from 'components/BMultiTab/BMultiTab';
import BondLogo from 'components/BondLogo';
import Loader from 'components/Loader';
import MenuMetric from 'components/Metrics/MenuMetric';
import { theme } from 'constants/theme';
import { selectFormattedReservePrice } from 'store/modules/app/app.selectors';
import { selectBondInstance, selectBondMetricsReady, selectBondPrice } from 'store/modules/bonds/bonds.selector';
import { Bond } from 'store/modules/bonds/bonds.types';
import { RootState } from 'store/store';

import { Mint } from './components/Mint';
import { Redeem } from './components/Redeem';

/**
 *
 * TODO :
 * - create mint view
 *  - add buttons
 *   + approve form
 *  - add custom slippage
 * - create redeem view
 *     - claim + claim and autostake
 */

const BondDetails = ({ open, bondID, bond }: { open: boolean; bondID: string; bond: Bond }) => {
    const history = useHistory();

    const onBackdropClick = () => history.goBack();

    const bashPrice = useSelector(selectFormattedReservePrice);
    const bondPrice = useSelector((state: RootState) => selectBondPrice(state, bondID));

    //TODO: Add the custom settings : Slippage & Recipient address

    const bondActions = [
        {
            label: t('mint:Mint'),
            component: <Mint bondID={bondID} />,
        },
        {
            label: t('mint:Redeem'),
            component: <Redeem bondID={bondID} />,
        },
    ];

    return (
        <Dialog
            {...{
                onBackdropClick,
                open,
                maxWidth: 'xl',
                fullWidth: true,
                PaperProps: { sx: { background: theme.palette.cardBackground.light, color: theme.palette.primary.dark } },
            }}
            sx={{
                p: 4,
                [theme.breakpoints.up('xs')]: {
                    p: 0,
                },
                backdropFilter: 'blur(10px)',
            }}
        >
            <DialogTitle sx={{ display: 'flex', alignItems: 'center', justifyContent: 'center', gap: theme.spacing(1) }}>
                <BondLogo bondLogoPath={bond.bondOptions.iconPath} isLP={bond.isLP()} />

                <Typography variant="body1">{bond.bondOptions.displayName}</Typography>
            </DialogTitle>
            <DialogContent>
                <Box>
                    <Grid container item xs={12} mb={4}>
                        <Grid item xs={12} sm={6}>
                            <MenuMetric key={'MintPrice'} metricKey={t('bond:MintPrice')} value={bondPrice} />
                        </Grid>

                        <Grid item xs={12} sm={6}>
                            <MenuMetric key={'BashPrice'} metricKey={t('BASHPrice')} value={bashPrice} />
                        </Grid>
                    </Grid>
                </Box>

                <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
                    <BMultiTabs tabs={bondActions} />
                </Box>
            </DialogContent>
        </Dialog>
    );
};

const BondDialogLoader = ({ open, bondID }: { open: boolean; bondID: string }) => {
    const metrics = useSelector((state: RootState) => selectBondMetricsReady(state, bondID));
    const bond = useSelector((state: RootState) => selectBondInstance(state, bondID));

    if (!metrics || !bond) return <Loader />;

    return <BondDetails open={open} bondID={bondID} bond={bond} />;
};

export default BondDialogLoader;
