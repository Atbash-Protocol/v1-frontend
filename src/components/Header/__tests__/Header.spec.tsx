import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import { BondDialog } from "components/BondDialog";
import { DEFAULT_NETWORK } from "constants/blockchain";
import { Web3Context } from "contexts/web3/web3.context";
import { BondType } from "helpers/bond/constants";
import { BondOptions } from "lib/bonds/bond/bond";
import { LPBond } from "lib/bonds/bond/lp-bond";
import { BondProviderEnum } from "lib/bonds/bonds.types";
import { Provider } from "react-redux";
import pendingTransactionsReducer from 'store/slices/pending-txns-slice';

import Header from "../index";


function renderComponent(component: JSX.Element, contextState?: any) {
    return render(
        <Web3Context.Provider value={contextState}>
            <Provider
                store={configureStore({
                    reducer: {
                        pendingTransactions: pendingTransactionsReducer,
                    },
                })}
            >
                {component}
            </Provider>
            ,
        </Web3Context.Provider>,
    );
}


describe('BondDialog', () => {
    const bondOptions: BondOptions = {
        name: 'bond',
        displayName: 'bondToDisplay',
        token: 'token',
        iconPath: 'path',
        lpProvider: BondProviderEnum.UNISWAP_V2,
        type: BondType.LP,
        networkID: 2,
        isActive: true,
    };

    const testBond = {
        bondInstance: new LPBond(bondOptions),
        metrics: {
            reserves: 0,
        },
        terms: {},
    };

    it('renders', () => {
        const { container } = renderComponent(<Header handleDrawerToggle={() => {}} isSmallScreen={false} />, {
            state: {
              networkID: DEFAULT_NETWORK
            },
            memoConnect: jest.fn(),
            memoDisconnect: jest.fn(),
        });
        expect(container).toMatchSnapshot();
    });


  })