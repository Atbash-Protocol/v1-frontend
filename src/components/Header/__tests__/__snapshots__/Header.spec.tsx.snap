// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BondDialog renders 1`] = `
<div>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionSticky css-13z6jsw-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-itmo3m-MuiToolbar-root"
      >
        <div
          class="MuiBox-root css-i9gxme"
        />
        <div
          class="MuiBox-root css-12z0wuy"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1carnhg-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Buy BASH
            </p>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-575ak1-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Disconnect
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
  </div>
  ,
</div>
`;
